### PR TITLE
pybind/rados: add list_lockers() & break_lock() & aio_unlock()

### PR DESCRIPTION
**Pull request by zjcmszh**
pybind/rados/rados.pyx: add list_lockers 

Signed-off-by: Zhang Jiao <zhangjiao_yewu@cmss.chinamobile.com>

### CODE REVIEW

**Comment by zjcmszh** [comment]
jenkins test docs

**Comment by djgalloway** [comment]
> @djgalloway please review, thanks.

Sorry, I'm not the best person to do a python code review; I was just trying to retrigger the `Docs: build check` check.

**Comment by tchaikov** [comment]
will review this changeset later this week.

**Comment by zjcmszh** [comment]
@djgalloway thanks very much

**Comment by zjcmszh** [comment]
@tchaikov  Really really thank you

File: src/pybind/rados/rados.pyx [hunk 1]
```diff
@@ -3869,6 +3869,66 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+                                c_tag[:tag_len].split(b'\0')]
+                     clients = [decode_cstr(client) for client in
+                                    c_clients[:clients_len].split(b'\0')]
+                     cookies = [decode_cstr(cookie) for cookie in
+                                    c_cookies[:cookies_len].split(b'\0')]
+                     addrs = [decode_cstr(client) for client in
+                                  c_addrs[:addrs_len].split(b'\0')]
+                     return list(zip(tag, clients, cookies, addrs))[:-1]
+                elif ret == 0:
+                    return None
```
Comment (tchaikov) [inline src/pybind/rados/rados.pyx:3954]: nit, IMHO, no need to make it a special case when there is no lockers, could just return an empty list, like
```py
return []
```

File: src/pybind/rados/rados.pyx [hunk 2]
```diff
@@ -3869,6 +3869,66 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+                elif ret == -errno.ERANGE:
+                    tag_len = tag_len * 2
+                    clients_len = clients_len * 2
+                    addrs_len = addrs_len * 2
```
Comment (tchaikov) [inline src/pybind/rados/rados.pyx:3958]: does it make sense to enlarge `cookies_len` as well?

File: src/pybind/rados/rados.pyx [hunk 3]
```diff
@@ -3311,6 +3311,41 @@ returned %d, but should return zero on success." % (self.name, ret))
             raise make_ex(ret, "aio_notify error: %s" % obj)
         return completion
 
+    def aio_unlock(self, key: str, name: str, cookie: str,
+                   oncomplete: Optional[Callable] = None) -> Completion:
```
Comment (tchaikov) [inline src/pybind/rados/rados.pyx:3315]: nit, s/oncomplete/on_complete/ for better readability.

File: src/test/pybind/test_rados.py [hunk 4]
```diff
@@ -1140,6 +1140,26 @@ def cb(blah):
...
+        count = [0]
+        def cb(blah):
+            with lock:
+                count[0] += 1
+                lock.notify()
+            return 0
+
+        self.ioctx.lock_exclusive("foo", "lock", "locker1", "desc_lock",
+                                  10000, 0)
+        comp = self.ioctx.aio_unlock("foo", "lock", "locker1", cb)
```
Comment (tchaikov) [inline src/test/pybind/test_rados.py:1154]: does it help if we could have a context manager for this pattern, like
```py
from contextlib import contextmanager
# ...
@contextmanager
def exclusive_lock(key: str, name: str, cookie: str, 
                   desc: str = "",
                   duration: Optional[int] = None,
                   flags: int = 0):
    def cb(_):
        return 0
    self.ioctx.lock_exclusive(key, name, cookie,
                              desc, duration, flags)
    try:
        yield
    finally:
        comp = self.ioctx.aio_unlock(key, name, cookie, cb)
        comp.wait_for_complete()

**Comment by tchaikov** [comment]
@zjcmszh ping?

**Pull request closed without merging**